src/_P084_VEML6070.ino

```diff
@@ -0,0 +1,230 @@
+        break;
+      }
+
+    case PLUGIN_GET_DEVICENAME:
+      {
+        string = F(PLUGIN_NAME_084);
+        break;
+      }
+
+    case PLUGIN_GET_DEVICEVALUENAMES:
+      {
+        strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_084));
+        strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[1], PSTR(PLUGIN_VALUENAME2_084));
+        strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[2], PSTR(PLUGIN_VALUENAME3_084));
+        break;
+      }
+
+    case PLUGIN_WEBFORM_LOAD:
+      {
+
+        String optionsMode[4] = { F("1/2T"), F("1T"), F("2T"), F("4T (Default)") };
+        addFormSelector(F("Refresh Time Determination"), F("itime"), 4, optionsMode, NULL, PCONFIG(1));
+
+        success = true;
+        break;
+      }
+
+    case PLUGIN_WEBFORM_SAVE:
+      {
+        //PCONFIG(0) = getFormItemInt(F("i2c_addr"));
+        PCONFIG(1) = getFormItemInt(F("itime"));
+
+        success = true;
+        break;
+      }
+
+    case PLUGIN_INIT:
+      {
+        bool error = VEML6070_Init(PCONFIG(1));
+        if(error){
+          String log = F("VEML6070: Not available!");
+          addLog(LOG_LEVEL_INFO, log);
+        }
+        
+        VEML6070_UvTableInit();
+
+        success = true;
+        break;
+      }
+
+    case PLUGIN_READ:
+      {
+        uint16_t uv_raw;
+        double uv_risk, uv_power;
+
+        uv_raw   = VEML6070_ReadUv();             // get UV raw values
+        uv_risk  = VEML6070_UvRiskLevel(uv_raw);  // get UV risk level
+        uv_power = VEML6070_UvPower(uv_risk);     // get UV power in W/m2
+
+        if (isnan(uv_raw) || uv_raw == (-1) || uv_raw == 65535) {
+          String log = F("VEML6070: no data read!");
+          addLog(LOG_LEVEL_INFO, log);
+          UserVar[event->BaseVarIndex + 0] = NAN;
+          UserVar[event->BaseVarIndex + 1] = NAN;
+          UserVar[event->BaseVarIndex + 2] = NAN;
+		      success = false;
+        } else {
+          UserVar[event->BaseVarIndex + 0] = uv_raw;
+          UserVar[event->BaseVarIndex + 1] = uv_risk;
+          UserVar[event->BaseVarIndex + 2] = uv_power;
+          String log = F("VEML6070: UV: ");
+          log += UserVar[event->BaseVarIndex];
+          addLog(LOG_LEVEL_INFO, log);
+		      success = true;
+        }
+
+        success = true;
+        break;
+      }
+
+  }
+  return success;
+}
+
+
+//////////////
+// VEML6070 //
+//////////////
+
+//get UV raw values
+uint16_t VEML6070_ReadUv(void)
+{
+  uint16_t uv_raw = 0;
+  // read high byte
+  if (Wire.requestFrom(VEML6070_ADDR_H, 1) != 1) {
+    return -1;
+  }
+  uv_raw   = Wire.read();
+  uv_raw <<= 8;
+  // read low byte
+  if (Wire.requestFrom(VEML6070_ADDR_L, 1) != 1) {
+    return -1;
+  }
+  uv_raw  |= Wire.read();
+  // high and low done
+  return uv_raw;
+
+}
+
+
+bool VEML6070_Init(byte it)
+{
+  byte error;
+  
+  Wire.begin();
+  Wire.beginTransmission(VEML6070_ADDR_L);
+  Wire.write((it << 2) | 0x02);
+  error = Wire.endTransmission();
+  delay(500);
```

Comment:
Do not use delay and especially not these long calls.
If there has to be some delay between calls, use the scheduler.
Delays other than a few msec are bad, since they're blocking the whole system.
So they should not be used in controllers or plugins.

The only places allowed to use delays are the scheduler and the setup function.

Reply:
Delay is removed. Left over from code where it was based on